doc/boxes.1.in
.\" @(#)boxes.1 2.3.2 10/03/2024
.\"
.TH boxes 1 "August 29 2026"
.UC 4
.SH NAME
boxes \- text mode box and comment drawing filter
.SH SYNOPSIS
.B boxes
[options] [infile [outfile]]
.SH DESCRIPTION
.I Boxes
is a text filter which can draw any kind of box around its input text. Box
design choices range from simple boxes to complex ASCII art. A box can also
be removed and repaired, even if it has been badly damaged by editing of the
text inside. Since boxes may be open on any side,
.I boxes
can also be used to create regional comments in any programming language.
New box designs can be added and shared by appending to a configuration file.
.LP
.I boxes
is a command line tool, but also integrates with any text editor that
supports filters. The
.I boxes
website has examples on how to configure editor integration for various
text editors:
.br
<URL:https://boxes.thomasjensen.com/editors.html>
.\" =======================================================================
.SH OPTIONS
Options offered by
.I boxes
are the following:
.TP 0.6i
\fB\-a\fP \fIformat\fP, \fB\-\-align\fP=\fIformat\fP
Alignment/positioning of text inside box. This option takes a format string
argument which is read from left to right. The format string may not
contain whitespace and must consist of one or more of the following
components:
.br

.I h\fPx
\- horizontal alignment of the input text block inside a box. Possible values
for
.I x
are
.I l
(ell, for left alignment),
.I c
(center), or
.I r
(right). This does not affect the justification of text lines within the
input text block (use the
.I j
argument instead).
.br
.I v\fPx
\- vertical alignment of the input text block inside a box. Possible values
for
.I x
are
.I t
(for top alignment),
.I c
(center), or
.I b
(bottom).
.br
.I j\fPx
\- justification of lines within the input text block. Possible values for
.I x
are
.I l
(ell, for left justification),
.I c
(center), or
.I r
(right). This does not affect the alignment of the input text block itself
within the box. Use the
.I h
and
.I v
arguments for input text block positioning.
.br

Short hand notations (can be combined with the above arguments):
.br
.I l
(ell) \- short for
.I h\fPl\fIv\fPc\fIj\fPl
.br
.I c
\- short for
.I h\fPc\fIv\fPc\fIj\fPc
.br
.I r
\- short for
.I h\fPr\fIv\fPc\fIj\fPr
.br

The default for
.B \-a
is
.I h\fPl\fIv\fPt.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP 0.6i
\fB\-c\fP \fIstring\fP, \fB\-\-create\fP=\fIstring\fP
Command line design definition for simple cases. The argument of this
option is the definition for the "west" (W) shape. The defined shape must
consist of exactly one line, i.e. no multi\-line shapes are allowed. The
.B \-c
option is intended as a shortcut for those cases where simple regional
comments are to be created, which only need a certain character or sequence
of characters to be placed in front of every line. In such cases, it is
much more convenient to simply specify
.B \-c
than to do a complete design definition in one's config file, where the
only shape defined is the west shape.
.br
This option implies a
.B \-d
and does not access the config file.
.B \-c
may of course be used in conjunction with any of the other options. By default,
.B \-c
is not specified.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP 0.6i
\fB\-\-color\fP, \fB\-\-no\-color\fP
Printing of color codes. Box designs and the text inside a box may contain ANSI
color codes, sometimes called "escape sequences". In this way, boxes and text
can be colored.
.br
Whether these escape sequences are printed by
.I boxes
is normally determined by the terminal capabilities (default). Using
\fB\-\-color\fP,
.I boxes
can be told to always output escape sequences even if it thinks the terminal
may not understand them. Using \fB\-\-no\-color\fP, escape sequences will
never be printed.

Of course, even with
.B \-\-color\fP,
a box will only appear colored if it is already defined with colors. In case
you want to auto-color some text that isn't yet, take a look at
.I lolcat\fP.
.br
These options consider all escape sequences to be color codes. Any other
escape sequences present will be printed or removed along with the color codes.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP 0.6i
\fB\-d\fP \fIstring\fP, \fB\-\-design\fP=\fIstring\fP
Design selection. The one argument of this option is the name of the design to
use, which may either be a design's primary name or any of its alias names.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP 0.6i
\fB\-e\fP \fIeol\fP, \fB\-\-eol\fP=\fIeol\fP
Override line terminator.
.I eol
can be
.I CR\fP,
.I LF\fP, or
.I CRLF\fP.
The default is to use the system-specific line terminator, which means
.I CRLF
on Windows, and
.I LF
otherwise. This option should only be used in an emergency, because normally
the system-specific line terminator will be just fine. This option is
considered experimental, and may go away in a future version of
.I boxes\fP.
Let us know in <URL:https://github.com/ascii-boxes/boxes/issues/60> if you
think we should keep it.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP 0.6i
\fB\-f\fP \fIstring\fP, \fB\-\-config\fP=\fIstring\fP
Use alternate config file. The one argument of this option is the name of a
valid
.I boxes
config file. The argument of
.B \-f
can also be a directory which contains a configuration file. More information
on this topic below in the CONFIGURATION FILE section.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP 0.6i
\fB\-h\fP, \fB\-\-help\fP
Print usage information.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP 0.6i
\fB\-i\fP \fIstring\fP, \fB\-\-indent\fP=\fIstring\fP
Indentation mode. Possible arguments are
.I text
(indent text inside of box),
.I box
(indent box, not text inside of box), or
.I none
(throw away indentation). Arguments may be abbreviated. The default is
.I box\fP.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP 0.6i
\fB\-k\fP \fIbool\fP, \fB\-\-kill\-blank\fP, \fB\-\-no\-kill\-blank\fP
Kill leading/trailing blank lines on removal. The value of
.I bool
is either
.I true
or
.I false\fP.
This option only takes effect in connection with
.B \-r\fP.
If set to
.I true\fP,
leading and trailing blank lines will be removed from the
output. If set to
.I false\fP,
the entire content of the former box is returned.
The default is
.I false\fP,
if both the top and the bottom part of the box are open, as is the case with
most regional comments. If the box's design defines a top part or a bottom
part, the default is
.I true\fP.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP 0.6i
\fB\-l\fP, \fB\-\-list\fP
List designs. Produces a listing of all available box designs in the
config file, along with a sample box and information about its creator.
Also checks the syntax of the entire config file, unless used in conjunction
with
.B \-d\fP,
which is to display detailed information about the specified design only.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP 0.6i
\fB\-m\fP, \fB\-\-mend\fP
Mend box. This removes a (potentially broken) box as with
.B \-r\fP,
and redraws it afterwards. The mended box is drawn according to the
options given. This may be important to know when it comes to restoring
padding, indentation, etc. for the mended box. Implies
.B \-k
.I false\fP.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP 0.6i
\fB\-n\fP \fIencoding\fP, \fB\-\-encoding\fP=\fIencoding\fP
Character encoding. Overrides the character encoding of the input and output
text. Choose from the list shown by \fIiconv -l\fP. If an invalid character
encoding is specified here, the system encoding is used as a fallback. The
default is also to use the system encoding, which is normally the best course
of action. So don't specify this option unless you must.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP 0.6i
\fB\-p\fP \fIstring\fP, \fB\-\-padding\fP=\fIstring\fP
Padding. Specify padding in spaces around the input text block for all
sides of the box. The argument string may not contain whitespace and must
consist of a combination of the following characters, each followed by a
number indicating the padding in spaces:
.br

.I a
\- (all) give padding for all sides at once
.br
.I h
\- (horiz) give padding for both horizontal sides
.br
.I v
\- (vertical) give padding for both vertical sides
.br
.I b
\- (bottom) give padding for bottom (south) side
.br
.I l
\- (left) give padding for left (west) side
.br
.I t
\- (top) give padding for top (north) side
.br
.I r
\- (right) give padding for right (east) side
.br

Example:
.B \-p
.I a\fP4\fIt\fP2
would define the padding to be 4 characters on all sides, except for the
top of the box, where the input text block will be only 2 lines away from
the box.
.br
The default for this option is determined by the box design used. If the
design does not specify anything, no default padding is used.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP 0.6i
\fB\-q\fP \fIquery\fP, \fB\-\-tag\-query\fP=\fIquery\fP
Query designs by tag. In contrast to
.B \-l\fP,
this will only print the matching design names. This option is normally used
stand-alone; if used in combination with other options, behavior is undefined.
The
.I query
argument is a comma-separated list of tags which can be present on a design
in order to match. A tag may optionally be prefixed with
.I \+
in order to require that it be present, or with
.I \-
in order to exclude designs which have that tag. Each tag can only occur once
per query. The special query
.I (all)
matches all box designs, even if they don't have any tags.
.br
This option is intended for use by scripts. Alias names are printed below
their primary design name, and postfixed with
.I (alias)\fP.
.br
Example:
.I boxes -q programming,-comment
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP 0.6i
\fB\-r\fP, \fB--remove\fP
Remove an existing box. Which design to use is detected automatically. In
order to save time or in case the detection does not decide correctly, combine
with
.B \-d
to specify the design.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP 0.6i
\fB\-s\fP \fIwidth\fP\fBx\fP\fIheight\fP, \fB\-\-size\fP=\fIwidth\fP\fBx\fP\fIheight\fP
Box size. This option specifies the desired box size in units of columns
(for width) and lines (for height).
If only a single number is given as argument, this number specifies the
desired box width. A single number prefixed by 'x' specifies only the box
height.  The actual resulting box size may vary depending on the individual
shape sizes of the chosen design. Also, other command line options may
influence the box size (such as
.B \-p\fP).
.br
By default, the smallest possible box is created around the text.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP 0.6i
\fB\-t\fP \fItabopts\fP, \fB\-\-tabs\fP=\fItabopts\fP
Tab handling. This option controls how tab characters in the input text are
handled. The
.I tabopts
must always begin with a positive integer number indicating the distance
between tab stops, sometimes called "spaces per tab".
.br
Immediately following the tab distance, an optional character can be appended,
telling
.I boxes
how to treat the leading tabs. The following options are available:
.br

.B e
\- expand tabs into spaces
.br
.B k
\- keep tabs as close to what they were as possible
.br
.B u
\- unexpand tabs. This makes
.I boxes
turn as many spaces as possible into tabs.
.br

The
.B \-t
.I string
can be just a number. In that case,
.B e
is assumed for tab handling. The factory default for the
.B \-t
option is simply 8, which is just such a case.
.br
For example, you could specify
.B \-t \fP4u
in order to have your leading tabs unexpanded. In the box content, tabs are
always converted into spaces. The tab distance in this example is 4.
.\" - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
.TP 0.6i
\fB\-v\fP, \fB\-\-version\fP
Print out current version number.
.\" =======================================================================
.SH CONFIGURATION FILE
.I Boxes
will look for the configuration file in several places, some of which are
given by the XDG specification.
.TP 0.6i
\fB1.\fP \fI\-f\fP option [file or dir]
When a configuration file is specified on the command line, we will use that. The
.B \-f
option can also specify a directory. Any location specified via
.B \-f
must exist, or
.I boxes
will terminate with an error.
.TP 0.6i
\fB2.\fP \fIBOXES\fP environment variable [file or dir]
If no config file is specified on the command line,
.I boxes
will check for the BOXES environment variable, which may contain a filename or
directory to use. Any location specified via the BOXES environment variable
must exist, or
.I boxes
will terminate with an error.
.TP 0.6i
\fB3.\fP \fI$HOME\fP [dir]
.TQ
\fB4.\fP \fI$XDG_CONFIG_HOME/boxes\fP [dir]
.TQ
\fB5.\fP \fI$HOME/.config/boxes\fP [dir]
.TQ
\fB6.\fP \fI--GLOBALCONF--\fP [file]
.TQ
\fB7.\fP \fI/etc/xdg/boxes\fP [dir]
.TQ
\fB8.\fP \fI/usr/local/share/boxes\fP [dir]
.TQ
\fB9.\fP \fI/usr/share/boxes\fP [dir]
Either one of these last two directory locations might have the same name as the
global config file from \fB6\fP. That's fine. It just means that we first
look for a file of that name, and then for a directory containing the file.
.P
The XDG environment variable \fIXDG_CONFIG_DIRS\fP is not supported. However,
its default value is supported via steps \fB8\fP and \fB9\fP above.
.TP 0.6i
In the above list, whenever a step is designated with [dir], the following file names will be found, in this order:
.br
\fB1.\fP .boxes
.br
\fB2.\fP box-designs
.br
\fB3.\fP boxes-config
.br
\fB4.\fP boxes
.LP
As soon as the first valid file is found, we use that and stop the search.
.P
The recommended location for a user-specific configuration file is
\fI$HOME/.boxes\fP or \fI$HOME/.config/boxes/.boxes\fP. A global
configuration file should be located at \fI--GLOBALCONF--\fP. But all of the
other locations are fully supported, too.
.P
The syntax of
.I boxes
config files is described on the website at
<URL:https://boxes.thomasjensen.com/config-syntax.html>.
.\" =======================================================================
.SH EXAMPLES
Examples on how to invoke
.I boxes
may be found on the website at
<URL:https://boxes.thomasjensen.com/examples.html>.
.br
Try
.P
\fIecho "Good Bye World!" | boxes -d nuke\fP
.P
.I Boxes
also combines nicely with other tools. Try
.P
\fIfiglet "boxes . . . !" | lolcat -f | boxes -d unicornthink\fP
.\" =======================================================================
.SH AVAILABILITY
The
.I boxes
website is <URL:https://boxes.thomasjensen.com/>. It contains a number
of examples illustrating this manual page as well as more in\-depth
documentation.
.\" =======================================================================
.SH AUTHOR
.I Boxes
was made by Thomas Jensen and the \fIboxes\fP contributors. It has been
lovingly maintained since 1999.
.br
For a full list of contributors, see
.br
<URL:https://boxes.thomasjensen.com/team.html#contributors>
.br
and <URL:https://github.com/ascii-boxes/boxes/graphs/contributors>.
.br
Please refer to the
.I boxes
website for the maintainer's current email address.
.\" =======================================================================
.SH VERSION
This is
.I boxes
version --BVERSION--.
.\" =======================================================================
.SH LICENSE
.I boxes
is free software under the terms of the GNU General Public License,
version 3. Details in the LICENSE file:
<URL:https://raw.githubusercontent.com/ascii-boxes/boxes/v--BVERSION--/LICENSE>
.\" =======================================================================
.SH ENVIRONMENT
.I Boxes
recognizes the following environment variables:
.HP 0.6i
BOXES
.br
Absolute path of the
.I boxes
configuration file. If this is specified, it must refer to an existing file
or directory.
.HP 0.6i
HOME
.br
The user's home directory.
.TP 0.6i
XDG_CONFIG_HOME
The root of the configuration file location as per the XDG specification.
.\" =======================================================================
.SH "SEE ALSO"
.BR figlet (6),
.BR iconv (1),
.BR lolcat (6)


pages/releases.md
---
title: Release History
permalink: /releases.html
redirect_from:
  - /docs/releases.html
created_at: 2015-03-17 22:06:12 +0100
last_modified_at: 2026-08-29 22:00:00 +0200
---

# Release Notes

Click on the version numbers to view the release notes for each release.

<table class="table table-striped">
  <thead>
    <tr>
      <th scope="col"><span class="d-none d-sm-inline">Release</span><span class="d-inline d-sm-none">Rel.</span></th>
      <th scope="col">Date</th>
      <th scope="col"><span class="d-none d-sm-inline">Sources</span><span class="d-inline d-sm-none">Src</span></th>
      <th scope="col"><span class="d-none d-sm-inline">Binaries</span><span class="d-inline d-sm-none">Bin</span></th>
      <th scope="col" class="d-none d-sm-table-cell">Remark</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th scope="row"><a href="{{ site.baseurl}}{% post_url 2026-08-29-boxes-v2.3.2-released %}">2.3.2</a></th>
      <td>August&nbsp;29, 2026</td>
      <td><a href="https://github.com/{{ site.github }}/tree/v2.3.2">link</a></td>
      <td><a href="https://github.com/{{ site.github }}/releases/tag/v2.3.2">link</a></td>
      <td class="d-none d-sm-table-cell">multiple fixes</td>
    </tr>
    <tr>
      <th scope="row"><a href="{{ site.baseurl}}{% post_url 2024-10-03-boxes-v2.3.1-released %}">2.3.1</a></th>
      <td>October&nbsp;3, 2024</td>
      <td><a href="https://github.com/{{ site.github }}/tree/v2.3.1">link</a></td>
      <td><a href="https://github.com/{{ site.github }}/releases/tag/v2.3.1">link</a></td>
      <td class="d-none d-sm-table-cell">a Windows fix</td>
    </tr>
    <tr>
      <th scope="row"><a href="{{ site.baseurl}}{% post_url 2024-02-17-boxes-v2.3.0-released %}">2.3.0</a></th>
      <td>February&nbsp;17, 2024</td>
      <td><a href="https://github.com/{{ site.github }}/tree/v2.3.0">link</a></td>
      <td><a href="https://github.com/{{ site.github }}/releases/tag/v2.3.0">link</a></td>
      <td class="d-none d-sm-table-cell">Colored UTF-8 boxes</td>
    </tr>
    <tr>
      <th scope="row"><a href="{{ site.baseurl}}{% post_url 2023-08-24-boxes-v2.2.1-released %}">2.2.1</a></th>
      <td>August&nbsp;24, 2023</td>
      <td><a href="https://github.com/{{ site.github }}/tree/v2.2.1">link</a></td>
      <td><a href="https://github.com/{{ site.github }}/releases/tag/v2.2.1">link</a></td>
      <td class="d-none d-sm-table-cell">multiple small fixes</td>
    </tr>
    <tr>
      <th scope="row"><a href="{{ site.baseurl}}{% post_url 2022-09-21-boxes-v2.2.0-released %}">2.2.0</a></th>
      <td>September&nbsp;21, 2022</td>
      <td><a href="https://github.com/{{ site.github }}/tree/v2.2.0">link</a></td>
      <td><a href="https://github.com/{{ site.github }}/releases/tag/v2.2.0">link</a></td>
      <td class="d-none d-sm-table-cell">license upgrade, cows</td>
    </tr>
    <tr>
      <th scope="row"><a href="{{ site.baseurl}}{% post_url 2021-06-14-boxes-v2.1.1-released %}">2.1.1</a></th>
      <td>June&nbsp;14, 2021</td>
      <td><a href="https://github.com/{{ site.github }}/tree/v2.1.1">link</a></td>
      <td><a href="https://github.com/{{ site.github }}/releases/tag/v2.1.1">link</a></td>
      <td class="d-none d-sm-table-cell">a few patches</td>
    </tr>
    <tr>
      <th scope="row"><a href="{{ site.baseurl}}{% post_url 2021-04-18-boxes-v2.1.0-released %}">2.1.0</a></th>
      <td>April&nbsp;18, 2021</td>
      <td><a href="https://github.com/{{ site.github }}/tree/v2.1.0">link</a></td>
      <td><a href="https://github.com/{{ site.github }}/releases/tag/v2.1.0">link</a></td>
      <td class="d-none d-sm-table-cell">many new features</td>
    </tr>
    <tr>
      <th scope="row"><a href="{{ site.baseurl}}{% post_url 2021-02-15-boxes-v2.0.0-released %}">2.0.0</a></th>
      <td>February&nbsp;15, 2021</td>
      <td><a href="https://github.com/{{ site.github }}/tree/v2.0.0">link</a></td>
      <td><a href="https://github.com/{{ site.github }}/releases/tag/v2.0.0">link</a></td>
      <td class="d-none d-sm-table-cell">Colored UTF-8 input</td>
    </tr>
    <tr>
      <th scope="row"><a href="{{ site.baseurl}}{% post_url 2019-02-21-boxes-v1.3-released %}">1.3</a></th>
      <td>February&nbsp;21, 2019</td>
      <td><a href="https://github.com/{{ site.github }}/tree/v1.3">link</a></td>
      <td><a href="https://github.com/{{ site.github }}/releases/tag/v1.3">link</a></td>
      <td class="d-none d-sm-table-cell">minor update</td>
    </tr>
    <tr>
      <th scope="row"><a href="{{ site.baseurl}}{% post_url 2017-03-16-boxes-v1.2-released %}">1.2</a></th>
      <td>March&nbsp;16, 2017</td>
      <td><a href="https://github.com/{{ site.github }}/tree/v1.2">link</a></td>
      <td><a href="https://github.com/{{ site.github }}/releases/tag/v1.2">link</a></td>
      <td class="d-none d-sm-table-cell">&nbsp;</td>
    </tr>
    <tr>
      <th scope="row"><a href="{{ site.baseurl}}{% post_url 2015-01-16-boxes-v1.1.2-released %}">1.1.2</a></th>
      <td>January&nbsp;16, 2015</td>
      <td><a href="https://github.com/{{ site.github }}/tree/v1.1.2">link</a></td>
      <td><a href="https://github.com/{{ site.github }}/releases/tag/v1.1.2">link</a></td>
      <td class="d-none d-sm-table-cell">&nbsp;</td>
    </tr>
    <tr>
      <th scope="row"><a href="https://github.com/{{ site.github }}/releases/tag/v1.1.1">1.1.1</a></th>
      <td>October&nbsp;19, 2012</td>
      <td><a href="https://github.com/{{ site.github }}/tree/release_1.1.1">link</a></td>
      <td><a href="https://github.com/{{ site.github }}/releases/tag/v1.1.1">link</a></td>
      <td class="d-none d-sm-table-cell">&nbsp;</td>
    </tr>
    <tr>
      <th scope="row"><a href="https://github.com/{{ site.github }}/releases/tag/v1.1">1.1</a></th>
      <td>July&nbsp;23, 2006</td>
      <td><a href="https://github.com/{{ site.github }}/tree/release_1.1">link</a></td>
      <td><a href="https://github.com/{{ site.github }}/releases/tag/v1.1">link</a></td>
      <td class="d-none d-sm-table-cell">new features</td>

doc/boxes.el
;;; boxes.el --- ASCII boxes unlimited! -*- lexical-binding: t -*-

;; Copyright (C) 1999, 2001, 2006 Jason L. Shiffer

;; Author: Jason L. Shiffer <jshiffer@zerotao.com>
;; Maintainer: Jason L. Shiffer <jshiffer@zerotao.com>
;; Version: 0.0
;; Package-Requires: ((emacs "24.3"))
;; Keywords: extensions
;; URL: https://boxes.thomasjensen.com
;; Created: 1999-10-30
;; Others:
;;   Vijay Lakshminarayanan: support for choosing boxes comment by current buffer mode.
;;   Mike Woolley: Customise and packaging support.

;; boxes - Command line filter to draw/remove ASCII boxes around text
;; Copyright (c) 1999-2024 Thomas Jensen and the boxes contributors
;;
;; This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public
;; License, version 3, as published by the Free Software Foundation.
;; This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied
;; warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
;; details.
;; You should have received a copy of the GNU General Public License along with this program.
;; If not, see <https://www.gnu.org/licenses/>.

;;; Commentary:

;; This program provides an interface to the boxes application which can be found at
;; https://boxes.thomasjensen.com/

;; To use this, add the following lines to your startup file:
;;
;;          (global-set-key "\C-cb" 'boxes-command-on-region)
;;          (global-set-key "\C-cq" 'boxes-create)
;;          (global-set-key "\C-cr" 'boxes-remove)
;;
;; If you didn't install the package version then you will also need to put this file somewhere in your load path and
;; add these lines to your startup file:
;;
;;          (autoload 'boxes-command-on-region "boxes" nil t)
;;          (autoload 'boxes-remove "boxes" nil t)
;;          (autoload 'boxes-create "boxes" nil t)
;;
;; If `use-package' is installed you can automatically install the boxes binary and elisp package with:
;;
;;          (use-package boxes
;;            :ensure t
;;            :ensure-system-package boxes
;;            :bind (("C-c b" . boxes-command-on-region)
;;                   ("C-c q" . boxes-create)
;;                   ("C-c r" . boxes-remove)))

;;; Code:

(defgroup boxes nil
  "ASCII boxes unlimited!"
  :prefix "boxes-"
  :group 'convenience)

(defcustom boxes-command "boxes"
  "The boxes command."
  :type 'string
  :group 'boxes)

(defcustom boxes-args ""
  "Additional arguments to the boxes command."
  :type 'string
  :group 'boxes)

(defcustom boxes-known-modes
  '((c-mode . "c-cmt2") (c++-mode . "c-cmt2") (java-mode . "java-cmt")
    (html-mode . "html-cmt") (sh-mode . "pound-cmt") (perl-mode . "pound-cmt")
    (python-mode . "pound-cmt") (ruby-mode . "pound-cmt")
    (emacs-lisp-mode . "lisp-cmt") (lisp-mode . "lisp-cmt"))
  "Default box type based on the major mode of the buffer."
  :type '(alist :key-type symbol :value-type string)
  :group 'boxes)

(defconst boxes-minimum-version "2.1.0"
  "Minimum required version of `boxes' to support querying the available box types.")

(defvar boxes-history nil
  "Boxes types history.")

(defvar-local boxes-default-type nil
  "The default type of box.")

(defvar boxes-types-list nil
  "List of types available to the current boxes implementation, nil if not set yet.")

(defun boxes-types ()
  "Return the list of types available to the current boxes implementation.
Signal error if a supported version of `boxes' is not available."
  (condition-case nil
      (or boxes-types-list
          (setq boxes-types-list
                (let ((types (process-lines boxes-command "-q" "(all)")))
                  (mapcar (lambda(type) (replace-regexp-in-string " *\(alias\) *$" "" type))
                          types))))
    (error (error "Please install Boxes %s or later" boxes-minimum-version))))

(defun boxes-default-type (mode)
  "Get the default box type for the given buffer major MODE."
  (or (cdr (assoc mode boxes-known-modes)) "c-cmt2"))

;;;###autoload
(defun boxes-create ()
  "Automagicly create a new box around the region based on the default type."
  (interactive "*")
  (boxes-command-on-region (region-beginning) (region-end) boxes-default-type))

;;;###autoload
(defun boxes-remove ()
  "Automagicly remove a new box around the region based on the default type."
  (interactive "*")
  (boxes-command-on-region (region-beginning) (region-end) boxes-default-type 1))

;;;###autoload
(defun boxes-command-on-region (start end type &optional remove)
  "Create or Remove boxes from a region.

To create a box select a region, hit \\[boxes-command-on-region]
& enter a box type.  Box type selection uses tab completion on
the supported types.

To remove a box simply prefix a 1 to the call, eg M-1
\\[boxes-command-on-region] will remove a box from a region.

Note that interactive use requires `boxes' >= 2.1.0 to support
querying the supported types.

When calling from Lisp, supply the region START & END and the box
TYPE to create a box.  Specifying a non-nil value for REMOVE,
removes the box."
  (interactive (let ((string
		      (completing-read (format "Box type (%s): " boxes-default-type)
				       (boxes-types) nil t nil 'boxes-history boxes-default-type)))
		(list (region-beginning) (region-end)
		      string
		      current-prefix-arg)))
  (when (or (null type) (string= type ""))
    (setq type (boxes-default-type major-mode)))
  (setq boxes-default-type type)
  (let ((command-string
	 (concat boxes-command
		 (if remove
		     (concat boxes-args " -r "))
		 (if type
		     (concat boxes-args " -d " type)))))
    (shell-command-on-region start end command-string nil 1)))

(provide 'boxes)
;;; boxes.el ends here
